File: app/src/main/java/xyz/rodit/snapmod/features/saving/ChatSaving.java
package xyz.rodit.snapmod.features.saving;

import android.content.Context;
import android.util.Log;

import java.io.IOException;
import java.lang.reflect.Proxy;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;
import xyz.rodit.snapmod.UriResolverSubscriber;
import xyz.rodit.snapmod.features.Feature;
import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.ChatActionHelper;
import xyz.rodit.snapmod.mappings.ChatMediaHandler;
import xyz.rodit.snapmod.mappings.ChatModelAudioNote;
import xyz.rodit.snapmod.mappings.ChatModelBase;
import xyz.rodit.snapmod.mappings.ChatModelLiveSnap;
import xyz.rodit.snapmod.mappings.ChatModelSavedSnap;
import xyz.rodit.snapmod.mappings.LiveSnapMedia;
import xyz.rodit.snapmod.mappings.MediaStreamProvider;
import xyz.rodit.snapmod.mappings.RxObserver;
import xyz.rodit.snapmod.mappings.SaveToCameraRollActionHandler;
import xyz.rodit.snapmod.mappings.SaveType;
import xyz.rodit.snapmod.util.PathManager;
import xyz.rodit.xposed.client.ConfigurationClient;
import xyz.rodit.xposed.client.FileClient;
import xyz.rodit.xposed.client.http.StreamProvider;
import xyz.rodit.xposed.client.http.StreamServer;
import xyz.rodit.xposed.client.http.streams.CachedStreamProvider;
import xyz.rodit.xposed.client.http.streams.FileProxyStreamProvider;

public class ChatSaving extends Feature {

    private final Map<Integer, Object> chatMediaMap = new HashMap<>();

    private ChatMediaHandler chatMediaHandler;

    public ChatSaving(FeatureContext context) {
        super(context);
    }

    @Override
    protected void performHooks() {
        // Obtain chat media handler instance
        ChatMediaHandler.constructors.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                chatMediaHandler = ChatMediaHandler.wrap(param.thisObject);
            }
        });

        // Allow save action in chat long-press menu.
        ChatActionHelper.canSave.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("allow_save_snaps")) {
                    param.setResult(true);
                }
            }
        });

        // Allow save action in chat model.
        ChatModelBase.canSave.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("allow_save_snaps")) {
                    param.setResult(true);
                }
            }
        });

        // Override save type to gallery to allow saving any snaps.
        ChatModelBase.getSaveType.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("allow_save_snaps")
                        && (ChatModelLiveSnap.isInstance(param.thisObject)
                        || ChatModelAudioNote.isInstance(param.thisObject))) {
                    param.setResult(SaveType.SNAPCHAT_ALBUM().instance);
                }
            }
        });

        // Map live snap model hashCode to media object for download later.
        ChatModelLiveSnap.constructors.hook(new XC_MethodHook() {
            @Override
            protected void afterHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("allow_save_snaps")) {
                    int hashCode = param.thisObject.hashCode();
                    chatMediaMap.put(hashCode, param.args[6]);
                }
            }
        });

        // Export non-savable media (live snaps and voice notes).
        SaveToCameraRollActionHandler.exportMedia.hook(new XC_MethodHook() {
            @Override
            protected void beforeHookedMethod(MethodHookParam param) {
                if (context.config.getBoolean("allow_save_snaps")) {
                    if (ChatModelLiveSnap.isInstance(param.args[1])) {
                        // Convert live snap to saved snap.
                        int hashCode = param.args[1].hashCode();
                        LiveSnapMedia media = LiveSnapMedia.wrap(chatMediaMap.get(hashCode));
                        ChatModelBase base = ChatModelBase.wrap(param.args[1]);
                        param.args[1] = new ChatModelSavedSnap(base.getContext(), base.getMessageData(), base.getSenderId(), Collections.emptyMap(), true, base.getReactionsViewModel(), true, 0, 0, media, null, base.getStatus(), true, true).instance;
                    } else if (ChatModelAudioNote.isInstance(param.args[1])) {
                        // Resolve audio uri and resolve through proxy of RxObserver.
                        // Note: the content resolver provided by appContext cannot open a stream from the uri.
                        ChatModelBase base = ChatModelBase.wrap(param.args[1]);
                        ChatModelAudioNote audio = ChatModelAudioNote.wrap(param.args[1]);
                        String dest = PathManager.getUri(context.config, PathManager.DOWNLOAD_AUDIO_NOTE, PathManager.createParams("id", base.getSenderId()), ".aac");
                        XposedBridge.log("Downloading audio note from " + audio.getUri() + " to " + dest + ".");
                        Object observerProxy = Proxy.newProxyInstance(context.classLoader, new Class[]{RxObserver.getMappedClass()}, new MediaUriDownloader(context.appContext, context.config, context.files, context.server, dest));
                        chatMediaHandler.resolve(audio.getUri(), Collections.emptySet(), true, Collections.emptySet()).subscribe(RxObserver.wrap(observerProxy));
                        param.setResult(null);
                    }
                }
            }
        });
    }

    private static class MediaUriDownloader extends UriResolverSubscriber {

        private static final String TAG = "MediaUriDownloader";

        public MediaUriDownloader(Context context, ConfigurationClient config, FileClient files, StreamServer server, String dest) {
            super(new UriListener(context, config, files, server, dest));
        }

        private static class UriListener implements ResolutionListener {

            private final Context context;
            private final ConfigurationClient config;
            private final FileClient files;
            private final StreamServer server;
            private final String dest;

            public UriListener(Context context, ConfigurationClient config, FileClient files, StreamServer server, String dest) {
                this.context = context;
                this.config = config;
                this.files = files;
                this.server = server;
                this.dest = dest;
            }

            @Override
            public void accept(Object result) {
                Log.d(TAG, "Accepted media stream provider: " + result);
                MediaStreamProvider streamProvider = MediaStreamProvider.wrap(result);
                String uuid = UUID.randomUUID().toString();
                StreamProvider provider = new CachedStreamProvider(new FileProxyStreamProvider(context, streamProvider::getMediaStream));
                try {
                    provider.provide();
                } catch (IOException e) {
                    Log.e(TAG, "Error pre-providing cached stream.", e);
                }

                server.mapStream(uuid, provider);
                files.download(config.getBoolean("use_android_download_manager", true), server.getRoot() + "/" + uuid, dest, "Audio Note", null);
            }
        }
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/saving/ChatSaving.kt
package xyz.rodit.snapmod.features.saving

import android.net.Uri
import xyz.rodit.snapmod.ResolutionListener
import xyz.rodit.snapmod.Shared
import xyz.rodit.snapmod.UriResolverSubscriber
import xyz.rodit.snapmod.features.Feature
import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.logging.log
import xyz.rodit.snapmod.mappings.*
import xyz.rodit.snapmod.util.PathManager
import xyz.rodit.snapmod.util.after
import xyz.rodit.snapmod.util.before
import xyz.rodit.snapmod.util.download
import xyz.rodit.xposed.client.http.StreamProvider
import xyz.rodit.xposed.client.http.streams.CachedStreamProvider
import xyz.rodit.xposed.client.http.streams.FileProxyStreamProvider
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.lang.reflect.Proxy

class ChatSaving(context: FeatureContext) : Feature(context) {

    private val chatMediaMap: MutableMap<Int, Any> = HashMap()

    private var chatMediaHandler: ChatMediaHandler? = null
    private var lastMessageData: MessageDataModel? = null

    override fun performHooks() {
        // Obtain chat media handler instance
        ChatMediaHandler.constructors.before {
            chatMediaHandler = ChatMediaHandler.wrap(it.thisObject)
        }

        // Allow save action in chat long-press menu.
        ChatActionHelper.canSave.before(context, "allow_save_snaps") {
            it.result = true
        }

        // Allow save action in chat model.
        ChatModelBase.canSave.before(context, "allow_save_snaps") {
            it.result = true
        }

        // Override save type to gallery to allow saving any snaps.
        ChatModelBase.getSaveType.before(context, "allow_save_snaps") {
            if (ChatModelLiveSnap.isInstance(it.thisObject)
                || ChatModelPlugin.isInstance(it.thisObject)
            ) {
                it.result = SaveType.SNAPCHAT_ALBUM().instance
            }
        }

        // Map live snap model hashCode to media object for download later.
        ChatModelWithMedia.constructors.after(context, "allow_save_snaps") {
            val hashCode = it.thisObject.hashCode()
            chatMediaMap[hashCode] = it.args[6]
        }

        // Export non-savable media (live snaps and voice notes).
        ActionMenuPresenter.handleAction.before {
            if (it.args[3] != ChatMenuItemType.SAVE_TO_CAMERA_ROLL().instance) return@before

            val base = ChatModelBase.wrap(it.args[2])
            lastMessageData = base.messageData

            when {
                ChatModelWithMedia.isInstance(it.args[2]) -> {
                    // Convert live snap to saved snap.
                    val hashCode = it.args[2].hashCode()
                    val media = LiveSnapMedia.wrap(chatMediaMap[hashCode])
                    it.args[2] = ChatModelSavedSnap(
                        base.context,
                        base.messageData,
                        base.senderId,
                        emptyMap<Any?, Any>(),
                        true,
                        null,
                        true,
                        0,
                        0,
                        media,
                        null,
                        base.status,
                        true
                    ).instance
                }
                ChatModelPlugin.isInstance(it.args[2]) -> {
                    val messageData = base.messageData

                    if (messageData.type != "audio_note") return@before
                    val media = GallerySnapMedia.wrap(messageData.media.instance).media
                    val uri = createMediaUri(messageData.arroyoMessageId, media.id)

                    resolveAndDownload(uri, messageData)
                    it.result = null
                }
            }
        }

        // Override snap save location by custom-downloading media instead.
        MediaExportControllerImpl.exportMedia.before(context, "enable_custom_snap_download_path") {
            val messageData = lastMessageData
            if (messageData?.isNotNull != true) return@before
            val destination = DestinationInfo.wrap(it.args[2])

            val provider = FileProxyStreamProvider(context.appContext) {
                FileInputStream(
                    destination.actualFileName?.let(::File)
                )
            }

            context.download(
                PathManager.DOWNLOAD_SNAP,
                createParams(messageData),
                '.' + destination.actualFileName.orEmpty().split('.').last(),
                provider,
                "${lastMessageData!!.senderDisplayName}'s snap"
            )

            it.result = null
        }
    }

    private fun createParams(messageData: MessageDataModel): Map<String, String> {
        return mapOf(
            "d" to messageData.senderDisplayName,
            "id" to messageData.senderId,
            "u" to messageData.senderUsername
        )
    }

    private fun resolveAndDownload(uri: Uri, messageData: MessageDataModel) {
        // Resolve audio uri and resolve through proxy of RxObserver.
        // Note: the content resolver provided by appContext cannot open a stream from the uri.
        val observerProxy = Proxy.newProxyInstance(
            context.classLoader,
            arrayOf(RxObserver.getMappedClass()),
            MediaUriDownloader(
                context,
                PathManager.DOWNLOAD_AUDIO_NOTE,
                createParams(messageData),
                ".aac"
            )
        )

        chatMediaHandler!!.resolve(
            uri,
            emptySet<Any>(),
            true,
            emptySet<Any>()
        ).subscribe(RxObserver.wrap(observerProxy))
    }

    private fun createMediaUri(messageId: String, mediaId: String): Uri {
        return Uri.Builder().scheme("content").authority("${Shared.SNAPCHAT_PACKAGE}.provider")
            .appendPath("chat_media")
            .appendPath(messageId)
            .appendPath(mediaId)
            .appendQueryParameter("target", "DEFAULT")
            .build()
    }

    private class MediaUriDownloader(
        context: FeatureContext, type: String, paramsMap: Map<String, String>, extension: String
    ) :
        UriResolverSubscriber(UriListener(context, type, paramsMap, extension)) {

        private class UriListener(
            private val context: FeatureContext, private val type: String,
            private val paramsMap: Map<String, String>, private val extension: String
        ) :
            ResolutionListener {

            override fun invoke(result: Any?) {
                log.debug("Accepted media stream provider: $result")
                val streamProvider = MediaStreamProvider.wrap(result)
                val provider: StreamProvider =
                    CachedStreamProvider(FileProxyStreamProvider(context.appContext) { streamProvider.mediaStream })
                try {
                    provider.provide()
                } catch (e: IOException) {
                    log.error("Error pre-providing cached stream.", e)
                }

                context.download(type, paramsMap, extension, provider, "Media")
            }
        }
    }
}